### PR TITLE
topdown: Insert serialized data in inter-query cache

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Previously when inserting an item in the inter-query cache, only
the size of the raw string representing the HTTP response message body
was used to calculate the amount of memory used by the item. This was
errornous as the actual item inserted in the cache would hold not only the
raw string but also the JSON representation of the message body.

The deserialized JSON body would end up consuming as much as 20x more memory compared
to the JSON encoded version of the same data. As a result, the cache would end
up consuming more memory than the user-defined cache size limit.

This change serializes the data that will be inserted in the inter-query cache
so as to get a more accurate representation of the size of the cache.

Fixes: #3042

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

**Review by tsandall — COMMENTED** [review]
These changes look fine. Let's take this opportunity to add some test coverage for the actual size of the elements inserted by http.send (I expected to see changes to the tests but there were none which indicates we were missing coverage in the first place.)

File: topdown/http.go [hunk 1]
```diff
@@ -679,63 +685,111 @@ func getBoolValFromReqObj(req ast.Object, key *ast.Term) (bool, error) {
...
+	return &interQueryCacheValue{Data: b}, nil
+}
+
+func (cb interQueryCacheValue) SizeInBytes() int64 {
+	return int64(len(cb.Data))
+}
+
+func (cb *interQueryCacheValue) getCacheData() (*interQueryCacheData, error) {
+	var res interQueryCacheData
+	err := json.Unmarshal(cb.Data, &res)
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply topdown/http.go]: Done.

File: topdown/http.go [hunk 2]
```diff
@@ -917,38 +971,44 @@ func parseMaxAgeCacheDirective(cc map[string]string) (deltaSeconds, error) {
...
+func prepareASTResult(headers http.Header, forceJSONDecode bool, body []byte, status string, statusCode int) (ast.Value, error) {
+	var resultBody interface{}
+
 	// If the response body cannot be JSON decoded,
 	// an error will not be returned. Instead the "body" field
 	// in the result will be null.
-	if isContentTypeJSON(resp.Header) || forceJSONDecode {
-		json.NewDecoder(&buf).Decode(&resultBody)
+	if isContentTypeJSON(headers) || forceJSONDecode {
+		json.Unmarshal(body, &resultBody)
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply topdown/http.go]: Done.

File: topdown/http.go [hunk 3]
```diff
@@ -565,66 +564,73 @@ func checkHTTPSendInterQueryCache(bctx BuiltinContext, key ast.Object, req *http
 		return nil, nil
 	}
 
+	cachedRespData, err := cachedResp.getCacheData()
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply topdown/http.go]: Done.

**Comment by ashutosh-narkar** [comment]
> These changes look fine. Let's take this opportunity to add some test coverage for the actual size of the elements inserted by http.send (I expected to see changes to the tests but there were none which indicates we were missing coverage in the first place.)

Added a test case for checking size of cache item.

**Review by tsandall — DISMISSED** [review]

**Pull request merged**